src/core.rs

```diff
@@ -338,7 +346,7 @@ pub struct ReedSolomon<F: Field> {
     parity_shard_count: usize,
     total_shard_count: usize,
     matrix: Matrix<F>,
-    tree: InversionTree<F>,
+    data_decode_matrix_cache: Mutex<LruCache<Vec<usize>, Arc<Matrix<F>>>>,
```

Comment:
Well, now I think we just want to make methods that require exclusive access take `&mut self` and let user decide if they want to wrap this in `Mutex` or not. Users may not need to pay for mutex overhead at all, but library forces them to.

This will be a breaking change of course.

Reply:
The counter argument is that looking up the data decode matrix from the cache takes only a fraction of the whole computation, and you only lock the mutex for that small fraction of time.

If the methods requires a `&mut self` and one has to wrap the entire `ReedSolomon` object in a `Mutex` then you will have to lock the mutex for the entire computation, which results in a lot more lock contention than the existing code.

Reply:
I see, this is just for decoding. Yeah, that'd be undesirable indeed.